src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/SpanNDHelpers.cs

```diff
@@ -0,0 +1,464 @@
+        {
+            nint totalLength = 1;
+            for (int i = 0; i < lengths.Length; i++)
+            {
+                totalLength *= lengths[i];
+            }
+
+            return totalLength;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static nint CalculateTotalLength(ReadOnlySpan<nint> lengths)
+        {
+            nint totalLength = 1;
+            for (int i = 0; i < lengths.Length; i++)
+            {
+                totalLength *= lengths[i];
+            }
+
+            return totalLength;
+        }
+
+        /// <summary>
+        /// Gets the set of strides that can be used to calculate the offset of n-dimensions in a 1-dimensional layout
+        /// </summary>
+        /// <returns></returns>
+        public static nint[] CalculateStrides(int rank, ReadOnlySpan<nint> lengths)
+        {
+            nint[] strides = new nint[rank];
+
+            nint stride = 1;
+
+            for (int i = strides.Length - 1; i >= 0; i--)
+            {
+                strides[i] = stride;
+                stride *= lengths[i];
+            }
+
+            return strides;
+        }
+
+        /// <summary>
+        /// Calculates the 1-d index for n-d indices in layout specified by strides.
+        /// </summary>
+        /// <param name="indices"></param>
+        /// <param name="strides"></param>
+        /// <param name="lengths"></param>
+        /// <returns></returns>
+        public static nint GetIndex(ReadOnlySpan<nint> indices, ReadOnlySpan<nint> strides, ReadOnlySpan<nint> lengths)
+        {
+            Debug.Assert(strides.Length == indices.Length);
+
+            nint index = 0;
+            for (int i = 0; i < indices.Length; i++)
+            {
+                if (indices[i] >= lengths[i])
+                    ThrowHelper.ThrowIndexOutOfRangeException();
+                index += strides[i] * (indices[i]);
+            }
+
+            return index;
+        }
+
+        public static void AdjustIndices(int curIndex, nint addend, ref nint[] curIndices, ReadOnlySpan<nint> lengths)
+        {
+            if (addend <= 0 || curIndex < 0)
+                return;
+            curIndices[curIndex] += addend;
+            AdjustIndices(curIndex - 1, curIndices[curIndex] / lengths[curIndex], ref curIndices, lengths);
+            curIndices[curIndex] = curIndices[curIndex] % lengths[curIndex];
+        }
+
+        public static void AdjustIndicesDown(int curIndex, nint addend, ref nint[] curIndices, ReadOnlySpan<nint> lengths)
+        {
+            if (addend <= 0 || curIndex < 0)
+                return;
+            curIndices[curIndex] -= addend;
+            if (curIndices[curIndex] < 0)
+            {
+                curIndices[curIndex] = lengths[curIndex] - 1;
+                AdjustIndices(curIndex - 1, 1, ref curIndices, lengths);
+            }
+        }
+
+        public static nint CountTrueElements(Tensor<bool> filter)
```

Comment:
Seems like this is mixing abstraction levels

Reply:
Sorry, I'm not sure I follow.

Reply:
Span implementation referencing Tensor

Reply:
Ah. So it was only being used by Tensors, but I put it in this file before I had the TensorHelpers file up. Moved this to where it belongs in that Tensors files.